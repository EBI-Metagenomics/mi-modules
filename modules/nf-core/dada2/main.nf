// TODO nf-core: If in doubt look at other nf-core/modules to see how we are doing things! :)
//               https://github.com/nf-core/modules/tree/master/modules/nf-core/
//               You can also ask for help via your pull request or on the #modules channel on the nf-core Slack workspace:
//               https://nf-co.re/join
// TODO nf-core: A module file SHOULD only define input and output files as command-line parameters.
//               All other parameters MUST be provided using the "task.ext" directive, see here:
//               https://www.nextflow.io/docs/latest/process.html#ext
//               where "task.ext" is a string.
//               Any parameters that need to be evaluated in the context of a particular sample
//               e.g. single-end/paired-end data MUST also be defined and evaluated appropriately.

process DADA2 {
    tag "$meta.id"
    label 'process_single'

    // TODO nf-core: List required Conda package(s).
    //               Software MUST be pinned to channel (i.e. "bioconda"), version (i.e. "1.10").
    // TODO nf-core: See section in main README for further information regarding finding and adding container addresses to the section below.
    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/bioconductor-dada2:1.30.0--r43hf17093f_0':
        'biocontainers/bioconductor-dada2:1.30.0--r43hf17093f_0' }

    input:
    tuple val(meta), path(reads)

    output:
    tuple val(meta), path("*asvs.fasta")    , emit: dada2_out
    path("*asv_counts.tsv")                 , emit: dada2_asv_counts
    path "versions.yml"                     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    if ( meta.single_end ){
        """
        $moduleDir/dada2.R \\
            $args \\
            $prefix \\
            $reads
        
        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            R: \$( R --version | head -1 | cut -d' ' -f3 )
            dada2: \$( R -e "suppressMessages(library(dada2));packageDescription('dada2')" | grep 'Version' | cut -d' ' -f2 )
        END_VERSIONS
        """
    } else {
        """
        $moduleDir/dada2.R \\
            $args \\
            $prefix \\
            ${reads[0]} \\
            ${reads[1]}

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            R: \$( R --version | head -1 | cut -d' ' -f3 )
            dada2: \$( R -e "suppressMessages(library(dada2));packageDescription('dada2')" | grep 'Version' | cut -d' ' -f2 )
        END_VERSIONS
        """
    }

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.asvs.fasta
    touch ${prefix}.asv_counts.tsv

    cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            R: \$( R --version | head -1 | cut -d' ' -f3 )
            dada2: \$( R -e "suppressMessages(library(dada2));packageDescription('dada2')" | grep 'Version' | cut -d' ' -f2 )
        END_VERSIONS
    """
}
