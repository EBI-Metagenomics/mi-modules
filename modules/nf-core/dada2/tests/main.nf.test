nextflow_process {

    name "Test Process DADA2"
    script "../main.nf"
    process "DADA2"

    tag "modules"
    tag "modules_nfcore"
    tag "dada2"

    test("Toy test fastq pared reads") {

        when {
            params {
                outdir = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file('${moduleDir}/tests/fixtures/test_noambig_1.fastq.gz', checkIfExists: true),
                        file('${moduleDir}/tests/fixtures/test_noambig_2.fastq.gz', checkIfExists: true) 
                    ]
                ]             
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.dada2_out).match() }
            )
        }

    }

    test("Toy test fastq pared reads - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    [
                        file('${projectDir}/modules/nf-core/dada2/tests/fixtures/test_noambig_1.fastq.gz', checkIfExists: true),
                        file('${projectDir}/modules/nf-core/dada2/tests/fixtures/test_noambig_2.fastq.gz', checkIfExists: true) 
                    ]
                ]        
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.dada2_out).match() }
            )
        }

    }

}
