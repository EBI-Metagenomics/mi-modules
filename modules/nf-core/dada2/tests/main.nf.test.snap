{
    "Toy test fastq single reads": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test_asvs.fasta:md5,f74d3b2d6572da28137bbd9fc782b0ac"
                    ]
                ],
                "1": [
                    "test_asv_counts.tsv:md5,acc9231178a33511c5bf9f087f8a586b"
                ],
                "2": [
                    "versions.yml:md5,686bb72a2339cb9bfa1f404c84266536"
                ],
                "dada2_asv_counts": [
                    "test_asv_counts.tsv:md5,acc9231178a33511c5bf9f087f8a586b"
                ],
                "dada2_out": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test_asvs.fasta:md5,f74d3b2d6572da28137bbd9fc782b0ac"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,686bb72a2339cb9bfa1f404c84266536"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-04T17:08:55.726691"
    },
    "Toy test fastq paired reads - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.asvs.fasta:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "test.asv_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "2": [
                    "versions.yml:md5,a431ec00e291eb67b51f237b42fcff00"
                ],
                "dada2_asv_counts": [
                    "test.asv_counts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ],
                "dada2_out": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test.asvs.fasta:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a431ec00e291eb67b51f237b42fcff00"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-04T17:09:03.860859"
    },
    "Toy test fastq paired reads": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_asvs.fasta:md5,7c861997aa34debd2ca2d80805c1cc25"
                    ]
                ],
                "1": [
                    "test_asv_counts.tsv:md5,e1744417a43b13a558ca0d93eb2b12fa"
                ],
                "2": [
                    "versions.yml:md5,686bb72a2339cb9bfa1f404c84266536"
                ],
                "dada2_asv_counts": [
                    "test_asv_counts.tsv:md5,e1744417a43b13a558ca0d93eb2b12fa"
                ],
                "dada2_out": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_asvs.fasta:md5,7c861997aa34debd2ca2d80805c1cc25"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,686bb72a2339cb9bfa1f404c84266536"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-09-04T17:08:41.196351"
    }
}